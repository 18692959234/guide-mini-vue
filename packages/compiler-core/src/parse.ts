import { NodeTypes } from "./ast";

type parseContext = {source: string}

const enum TagTypes {
  Start,
  End
}

export function baseParse (content) {
  const context: parseContext = createParserContext(content);
  return createRoot(parseChildren(context, []));
}

function parseChildren (context, ancestors) {
  const nodes:any = [];
  while(!isEnd(context, ancestors)) {
    let node;
    const s = context.source;
    if (s.startsWith('{{')) {
      node = parseInterpolation(context);
    } else if (s[0] === '<') {
      if (/[a-z]/i.test(s[1])) {
        node = parseElement(context, ancestors);
      }
    }
  
    if (!node) {
      node = parseText(context);
    }
    nodes.push(node)
  }

  return nodes;
}

function isEnd (context, ancestors) {
  const s = context.source;
  if (s.startsWith('</')) {
    for (let i = ancestors.length - 1; i >= 0; i--) {
      const tag = ancestors[i].tag;
      if (startsWithEndTagOpen(s, tag)) {
        return true;
      }
    }
  }

  return !s;
}

function parseInterpolation (context) {

  const openDelimiter = '{{';
  const closeDelimiter = '}}';
  const openDelimiterLen = openDelimiter.length;
  const closeDelimiterLen = closeDelimiter.length;

  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiterLen);
  
  advanceBy(context, openDelimiterLen)

  const rawContentLength = closeIndex - openDelimiterLen;

  const rawContent = parseTextData(context, rawContentLength);

  const content = rawContent.trim()
  // {{message}} -> message
  advanceBy(context, closeDelimiterLen);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  }
}

function advanceBy (context, len: number) {
  context.source = context.source.slice(len)
}

function createRoot (children) {
  return {
    children,
    type: NodeTypes.ROOT
  }
}

function createParserContext (content) {
  return {
    source: content
  }
}

function parseElement(context, ancestors): any {
  

  const element: any = parseTag(context, TagTypes.Start);
  
  ancestors.push(element);
  element.children = parseChildren(context, ancestors);
  ancestors.pop();
  console.log('-----',context.source,  element.tag, ancestors)
  if (startsWithEndTagOpen(context.source, element.tag)) {
    parseTag(context, TagTypes.End);
  } else {
    throw new Error(`缺少结束标签:${element.tag}`);
  }
  return element;
}

function startsWithEndTagOpen(source, tag) {
  return (
    source.startsWith("</") &&
    source.slice(2, 2 + tag.length).toLowerCase() === tag.toLowerCase()
  );
}


function parseTag (context, type: TagTypes) {
  const match:any = /\<\/?([a-z]*)/.exec(context.source);
  const tag = match[1];
  advanceBy(context, match[0].length)
  advanceBy(context, 1);
  if (type === TagTypes.End) return;
  return {
    type: NodeTypes.ELEMENT,
    tag,
  }
}

function parseText(context: any): any {
  const endTokens = ['<', "{{",];
  const s = context.source;
  let endIndex = s.length;
  for (let i = 0; i < endTokens.length; i++) {
    const index = s.indexOf(endTokens[i]);
    if (index !== -1 && endIndex > index) {
      endIndex = index;
    }
  }
  const content = parseTextData(context, endIndex);
  return {
    type: NodeTypes.TEXT,
    content,
  }
}

function parseTextData (context, length) {
  const content = context.source.slice(0, length);
  advanceBy(context, length);
  return content
}
