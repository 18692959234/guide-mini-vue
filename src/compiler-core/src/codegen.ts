export function generate(ast) {
  const context = createCodegenContext();
  const { push } = context;
  push('return ');
  
  const functionName = 'render';
  const args = ['_ctx', '_cache'];
  const signature = args.join(', ');
  push(`function ${functionName} (${signature}) {`);
  genNode(ast, context);
  push(`}`);
  return {code: context.code};
}

function genNode (ast, context) {
  const node = ast.codegenNode;
  context.push(`return '${node.content}'`);
  return context.code;
}

function createCodegenContext () {
  const context = {
    code: '',
    push (source) {
      context.code += source;
    }
  }
  return context;
}