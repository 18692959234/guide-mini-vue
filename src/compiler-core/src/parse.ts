import { NodeTypes } from "./ast";

type parseContext = {source: string}

export function baseParse (content) {
  const context: parseContext = createParserContext(content);
  return createRoot(parseChildren(context));
}

function parseChildren (context) {
  const nodes:any = [];

  let node;

  if (context.source.startsWith('{{')) {
    node = parseInterpolation(context);
  }

  nodes.push(node);

  return nodes;
}

function parseInterpolation (context) {

  const openDelimiter = '{{';
  const closeDelimiter = '}}';
  const openDelimiterLen = openDelimiter.length;
  const closeDelimiterLen = closeDelimiter.length;

  const closeIndex = context.source.indexOf(closeDelimiter, openDelimiterLen);
  
  advanceBy(context, openDelimiterLen)

  const rawContentLength = closeIndex - openDelimiterLen;

  const rawContent = context.source.slice(0, rawContentLength);

  const content = rawContent.trim()
  // {{message}} -> message
  advanceBy(context, rawContentLength + closeDelimiterLen);

  return {
    type: NodeTypes.INTERPOLATION,
    content: {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content,
    },
  }
}

function advanceBy (context, len: number) {
  context.source = context.source.slice(len)
}

function createRoot (children) {
  return {
    children
  }
}

function createParserContext (content) {
  return {
    source: content
  }
}