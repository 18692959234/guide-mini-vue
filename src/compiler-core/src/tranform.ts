import { NodeTypes } from "./ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform (root, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createCodegenNode(root);
  root.helpers = [...context.helpers.keys()];
}

function createTransformContext(root: any, options: any) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper (key) {
      context.helpers.set(key, 1)
    }
  };
  return context;
}

function traverseNode(node: any, context: any) {
  const nodeTransforms = context.nodeTransforms;
  const exitFns: any = [];
  if (nodeTransforms.length > 0) {
    for (let index = 0; index < nodeTransforms.length; index++) {
      const plugin = nodeTransforms[index];
      const onExit  = plugin(node, context);
      if (onExit) exitFns.push(onExit);
    }
  }

  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      const { helper } = context;
      helper(TO_DISPLAY_STRING);
      break;
    case NodeTypes.ELEMENT:
    case NodeTypes.ROOT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }
  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}


function traverseChildren (node, context) {
  const children = node.children;
  for (let index = 0; index < children.length; index++) {
    const transform = children[index];
    traverseNode(transform, context)
  }
}
function createCodegenNode(root: any) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode
  } else {
    root.codegenNode = root.children[0];
  }
}

