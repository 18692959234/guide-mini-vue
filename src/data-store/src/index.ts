export function createDataStore (data) {
  return defineReactive(data);
}

function defineReactive(data) {
  return new Proxy(data, Object.assign({}, {get: createGetter(), set: createSetter()}))
}


function createGetter () {
  return function get (target, key) {
    const res = Reflect.get(target, key);
    if (isShouldTrack()) {
      track(target, key)
    }
    return res;
  }
}

const targetDepMap = new Map();

function getDep (target, key) {
  let depsMap = targetDepMap.get(target);
  if (!depsMap) {
    depsMap = new Map();
    targetDepMap.set(target, depsMap);
  }
  let dep = depsMap.get(key);
  if (!dep) {
    dep = new Set();
    depsMap.set(key, dep)
  }
  return dep;
}

function track (target, key) {
  const dep = getDep(target, key);
  dep.forEach(d => d.run());
}

function trigger (target, key) {
  const dep = getDep(target, key);
  if (!dep.has(activeWatchEffect)) {
    dep.add(activeWatchEffect);
  }
}

function createSetter () {
  return function set (target, key, val) {
    const res = Reflect.set(target, key, val);
    trigger(target, key);
    return res;
  }
}

let activeWatchEffect;

class WatchEffect {
  private _fn;
  constructor (fn) {
    this._fn = fn;
  }

  run () {
    activeWatchEffect = this;
    this._fn();
    activeWatchEffect = null;
  }
}

function isShouldTrack () {
  return !!activeWatchEffect;
}

export function watch (watcher, fn) {
  const effect = new WatchEffect(fn);
}
