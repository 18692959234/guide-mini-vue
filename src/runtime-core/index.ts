export { h } from './h'

export { renderSlots } from './helpers/renderSlots';

export { createTextVNode } from './vnode';

export { getCurrentInstance } from './components'

export { provide, inject } from './apiInject';

export { createRenderer } from './renderer'

// 流程
// 调用createApp初始化 返回一个对象
// 执行mount方法
// mount方法里首先 拿到根实例去初始化一个vnode， 然后调用render
// render里直接调用patch
// patch里面需要判断vnode是否为组件或者是element 分别调用processComponent和processElement 去走一遍流程
// processComponent里面需要去挂载组件既mountComponent
// mountComponent 首先创建一个组件实例，然后去初始化组件既： setupComponent
// setupComponent需要做的事情 1、初始化props 2、 初始化 slot 3、初始化有状态的组件 setupfulStateComponent
// setupfulStateComponent 里主要是把有setup与render的vnode 给绑定到组件实例上
// 最后执行setupRenderEffect 主要是把调用render 并把render返回的结果丢给patch 再次执行
