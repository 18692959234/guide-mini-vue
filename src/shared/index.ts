export * from './toDisplayString'

export const extend = Object.assign;

export const EMPTY_OBJ = {};

export const isObject = (value) => value != null && typeof value === 'object';

export const isString = (value) => typeof value === 'string'

export const hasChanged = (value, newValue) => !Object.is(value, newValue);

export const hasOwn = (val, key) =>
  Object.prototype.hasOwnProperty.call(val, key);

/**
 * @name 转驼峰方法
 * @param {String} str 
 * @returns {String} res
 */
export const camelize = (str: string) => {
  return str.replace(/-(\w)/g, (_, c: string) => {
    return c ? c.toUpperCase() : "";
  });
};

const capitalize = (str: string) => {
  return str.charAt(0).toUpperCase() + str.slice(1);
};

/**
 * @name 处理on函数
 * @param str 
 * @returns {String} res
 */
export const toHandlerKey = (str: string) => {
  return str ? "on" + capitalize(str) : "";
};

export const clog = (...arg) => console.log('--------', ...arg, '--------')