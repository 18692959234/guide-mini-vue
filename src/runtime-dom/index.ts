import { createRenderer } from "../runtime-core/renderer";

function createElement (type) {
  return document.createElement(type);
}

function patchProp (el, key, prevValue, nextValue) {
    const isOn = (key) => /^on[A-Z]/.test(key)
    console.log(prevValue, nextValue)
    if (isOn(key)) {
      const event = key.slice(2).toLowerCase();
      el.addEventListener(event, nextValue);
    } else {
      console.log(nextValue != null)
      if (nextValue != null) {
        el.setAttribute(key, nextValue)
      } else {
        el.removeAttribute(key);
      }
    }
}

function insert (el, parent) {
  parent.append(el);
}

const renderer:any = createRenderer({
  createElement,
  patchProp,
  insert
})

export function createApp (...args) {
  return renderer.createApp(...args)
}

export * from '../runtime-core'