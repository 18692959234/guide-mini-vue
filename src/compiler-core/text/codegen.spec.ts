import { describe, expect, it, test } from 'vitest';
import { generate } from '../src/codegen';
import { baseParse } from '../src/parse';
import { transform } from '../src/tranform';
import { transformExpression } from '../src/transform/transformExpression';

describe('codegen', () => {
  it('string', () => {
    const ast = baseParse('hi');
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  })
})

describe('codegen', () => {
  it('interpolation', () => {
    const ast = baseParse('{{message}}');
    transform(ast, {
      nodeTransforms: [transformExpression]
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  })
})