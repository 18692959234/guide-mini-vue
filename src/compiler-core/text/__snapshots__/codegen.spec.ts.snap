// Vitest Snapshot v1

exports[`codegen element > element 1`] = `
"const { toDisplayString:_toDisplayString, createElementVNode:_createElementVNode } = vue
return function render (_ctx, _cache) {return _createElementVNode('div', null, 'hi,' + _toDisplayString(_ctx.message))}"
`;
